# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     |  Julieta Madrigal Flores |
| **Fecha**      | Lunes 16 de marzo del 2026  |
| **Expediente** |  744029 |

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion).

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$.

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds.

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [ ]:
import pandas as pd
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, KFold

##**1. Modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid')**

In [ ]:
from sklearn.datasets import load_iris

iris = load_iris()

# .data son las features (las medidas de las flores)
X_iris = iris.data

# .target es lo que queremos predecir (la clase de la flor)
y_iris = iris.target

print(iris.target_names)

['setosa' 'versicolor' 'virginica']


In [ ]:
print(X_iris.shape)

(150, 4)


In [ ]:
kernels_iris  = ['linear', 'poly', 'rbf', 'sigmoid']
for kernel in kernels_iris:
  # NO ponemos el max_iter porque el dataset es muy chiquito y no se va a tardar como el de default
    modelo_iris = SVC(kernel=kernel, probability=True)

## **2.Modela con LogisticRegression**


In [ ]:
lr_iris = LogisticRegression()

##**3.Utiliza el AUC como métrico de Cross-Validation**

In [ ]:
kf_iris = KFold(n_splits=10, shuffle=True, random_state=42)
resultados_iris = {}

for kernel in kernels_iris:
    modelo_iris = SVC(kernel=kernel, probability=True)
    # NOTAAAAAAAAAA
    # roc_auc — funciona solo para problemas binarios (0 o 1) como Default donde solo hay DOS (1,0) clases
    # roc_auc_ovr — funciona para problemas multiclasecomo Iris donde hay 3 clases
    # convierte el problema en varios problemas binarios (One vs Rest) para poder calcular el AUC
    scores = cross_val_score(modelo_iris, X_iris, y_iris, cv=kf_iris, scoring='roc_auc_ovr')
    resultados_iris[kernel] = scores.mean()

scores_lr_iris = cross_val_score(lr_iris, X_iris, y_iris, cv=kf_iris, scoring='roc_auc_ovr')
resultados_iris['Logistic Regression'] = scores_lr_iris.mean()

##**4.Compara resultados**

In [ ]:
pd.DataFrame(resultados_iris, index=['AUC']).T

,AUC
linear,1.000000
poly,0.998625
rbf,0.997413
sigmoid,0.739585
Logistic Regression,0.997413


##**INTERPRETACIÓN**

A partir de los resultados obtenidos se puede concluir que el mejor modelo y con un perfecto AUC de 1.0 fue el kernel linear, sin embargo este resultado es muy sospecho o muy 'perfecto' ya que podría indicar sobreajuste. Esto puede ser porque el dataset de Iris es muy pequeño, solo tiene 150 filas y las clases fueron muy faciles de separar, por lo que puede que el modelo memorizara datos en lugar de solo aprender los patrones generales. Asimismo, poly, rbf y logistic regression obtuvieron resultados casi perfectos de 0.9986 y 0.9974, lo que sugiere que el dataset fue muy fácil de clasificar. Sigmoid fue el único modelo que obtuvo un resultado inferior con 0.7385, lo que indica que este kernel no es el más adecuado para este tipo de datos.

## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [ ]:
data = pd.read_csv('Default.csv')
data.head()

,default,student,balance,income
0,No,No,729.526495,44361.625074
1,No,Yes,817.180407,12106.134700
2,No,No,1073.549164,31767.138950
3,No,No,529.250605,35704.493940
4,No,No,785.655883,38463.495880


In [ ]:
# La variables categóricas, las convertimos en boleanas
data['default'] = data['default'].map({'No': 0, 'Yes': 1})
data['student'] = data['student'].map({'No': 0, 'Yes': 1})
data.head()

,default,student,balance,income
0,0,0,729.526495,44361.625074
1,0,1,817.180407,12106.134700
2,0,0,1073.549164,31767.138950
3,0,0,529.250605,35704.493940
4,0,0,785.655883,38463.495880


In [ ]:
# Definir X y Y
X = data.drop('default', axis=1)
y = data['default']

##**1. Modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid')**

In [ ]:
# SVC (Support Vector Classifier) = busca el hiperplano que mejor separa las clases
#  Cada kernel transforma los datos de una manera diferente para encontrar ese hiperplano
# linear — separación con una línea recta
# poly — separación con una curva polinomial
# rbf — separación con una función gaussiana (el más usado)
# sigmoid — separación similar a una red neuronal

# kernels = crea lista con los 4 kernels que pide el lab
# for kernel in kernels = itera sobre cada kernel
# SVC(kernel=kernel, probability=True) = crea un modelo SVC con el kernel de turno
# probability=True = calcula la probabilidad de que cada dato pertenezca a cada clase y después sirve para calcular AUC


kernels = ['linear', 'poly', 'rbf', 'sigmoid']
# Realmente podríamos dejar solo la creación de la lista porque en el paso 3 (es donde todo se ejecuta)
# vamos a tener que volver a hacer este for pero con más cosas


for kernel in kernels:
    modelo = SVC(kernel=kernel, probability=True)

# Ahorita se define cómo se va a comportar cada modelo según su kernel pero todavía NO está entrenado NI se está evaluando

## **2.Modela con LogisticRegression**

In [ ]:
# solo definimos NO ejecutamos
lr = LogisticRegression()

##**3.Utiliza el AUC como métrico de Cross-Validation**

In [ ]:
# n_splits=10 divide el dataset en 10 partes = 9 partes se usan para train y 1 para test
# shuffle=True = mezcla los datos antes de dividir para que las particiones no queden sesgadas por el orden original
kf = KFold(n_splits=10, shuffle=True, random_state=42)

resultados = {}
# Crea un diccionario vacío donde vamos a guardar los resultados de cada modelo
# RECUERDA que una lista = [0.6409, 0.8529, 0.6345, 0.5373] y un diccionario = {'linear': 0.6409, 'poly': 0.8529, 'rbf': 0.6345, 'sigmoid': 0.5373}


# Por cada kernel, se crea un modelo, se evalúa con el K-Fold y guarda el AUC promedio en el diccionario
for kernel in kernels:
    # ponemos max_iter=1000 porque cuando lo hice sin límite de iteraciones tardaba demasiado
    # Con max_iter=1000 limitamos las iteraciones para que corra más rápido
    modelo = SVC(kernel=kernel, probability=True, max_iter=1000)

    # Entrena y evalúa el modelo 10 veces (una por cada partición del K-Fold)
    # scores es un arreglo con 10 AUCs, uno por cada partición
    # osea hace 10 veces el modelo "poly" porque debe de hacerlo con cada parte
    # AQUI ES DONDE DECIDES QUE MÉTRICO USAR = SOLO CAMBIAS LO QUE ESTÁ EN SCORING
    scores = cross_val_score(modelo, X, y, cv=kf, scoring='roc_auc')

    # Saca el promedio de los 10 AUCs y lo guarda en el diccionario
    resultados[kernel] = scores.mean()

# hace Logistic Regression 10 veces = una por cada partición del K-Fold
# aquí NOOO tiene kernel, es la regresión de la pura partición SIN KERNEL
scores_lr = cross_val_score(lr, X, y, cv=kf, scoring='roc_auc')
resultados['Logistic Regression'] = scores_lr.mean()

/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider pre-processing your data with StandardScaler or MinMaxScaler.
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn/svm/_base.py:305: ConvergenceWarning: Solver terminated early (max_iter=1000).  Consider

##**4.Compara resultados**

In [ ]:
pd.DataFrame(resultados, index=['AUC']).T

# ROC es una curva que mide qué tan bien el modelo separa las clases y AUC es el área bajo esa curva

# AUC = 1 → modelo perfecto
# AUC = 0.5 → modelo adivina al azar
# AUC = 0 → modelo siempre se equivoca

,AUC
linear,0.640940
poly,0.852937
rbf,0.634505
sigmoid,0.537251
Logistic Regression,0.949096


##**INTERPRETACIÓN**

A partir de los resultados obtenidos, se puede concluir que Logistic Regression fue el mejor modelo con un AUC de 0.9491, esto tiene sentido porque default es una variable binaria y Logistic Regression predice una probabilidad entre 0 y 1; es decir, Logistic Regression funciona muy bien porque justo fue creada para este tipo de problemas. Ahora bien, en los kernel, poly fue el mejor con 0.8529, demostrando que una transformación polinomial logra separar bien las clases; mientras que linear y rbf obtuvieron 0.6409 y 0.6345, resultados bastante malos, indicando que una separación linear y una gaussiana NO son adecuadas para estos datos. Finalmente, el peor fue sigmoid con 0.5373 = el modelo similar a una red neuronal está casi casi adivinando al alzar como separar el dataset.

# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

#**NOTA**

Si quisiera hacer alguno de los métricos disponibles en el addendum solo tendrías que cambiar el parámetro scoring en cross_val_score

En lugar de 'roc_auc', pones cualquier métrico del addendum:

- scores = cross_val_score(modelo, X, y, cv=kf, scoring='accuracy')
- scores = cross_val_score(modelo, X, y, cv=kf, scoring='f1')
- scores = cross_val_score(modelo, X, y, cv=kf, scoring='precision')

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result